### ISSUE TITLE
The specified module could not be found, with electron v4.0.0.

### ISSUE THREAD

**Issue by iacovlev-pavel** [lovell/sharp#1522]
With electron 3.0.12 everything seems to work fine.
Changed version to 4.0.0, removed node_modules, did `npm i`, all ok.

If I run electron-rebuild, the command runs, but it does not rebuild the sharp module, does not seem it's picking it up.
If I run `require('sharp')` I get:
`The specified module could not be found`

If I cd to node_modules/sharp and try to recompile it manually:
`node-gyp rebuild --target=4.0.0 --arch=x64 --dist-url=https://atom.io/download/electron`
and try `require('sharp')` I still get `The specified module could not be found`.

Note: I have checked and `sharp.node` is present at the specified path.

**Comment by lovell** [comment]
The `target` flag passed to node-gyp refers to the version of Node rather than the version of Electron.

You'll probably want to use [electron-rebuild](https://github.com/electron/electron-rebuild).

**Comment by iacovlev-pavel** [comment]
Hello, electron-rebuild is what I am current using with Electron 3, but it fails with Electron 4.

From what I understand the version is for electron since `dist-url` is set to `https://atom.io/download/electron`

**Comment by ddding** [comment]
[sharp.zip](https://github.com/lovell/sharp/files/2716240/sharp.zip)
I have the same issue and here is my Dependency Walker file.

**Comment by lovell** [comment]
Thanks, according to that `.dwi` file the missing shared library is `node.dll` so I guess either something changed in the way Electron 4 provides/packages this and/or electron-rebuild does not yet support it.

I'd imagine other native modules are going to fail in a similar manner - please can you report this upstream.

**Comment by lovell** [comment]
> "However, in Electron 4.x and higher, the symbols needed by native modules are exported by electron.exe, and there is no node.dll in Electron 4.x."

https://electronjs.org/docs/tutorial/using-native-node-modules#troubleshooting

**Comment by ddding** [comment]
Thank you, but I still don't understand how to make things right, or maybe I should just wait for electron-rebuild to release some update?

**Comment by lovell** [comment]
Please can someone confirm that this problem is limited to Windows only. Is anyone able to produce a minimal repo that exhibits the problem?

**Comment by groboter** [comment]
I have placed a very simple repo which exhibits the problem here [https://github.com/groboter/sharpErrorTestApp](https://github.com/groboter/sharpErrorTestApp). I hope it is ok, because it is my first one! 

Basically it is just the Quick-Start-App from [https://github.com/electron/electron-quick-start](https://github.com/electron/electron-quick-start) and I just added line 3 `const {sharp} = require('sharp')` which produces the error.

**Comment by lovell** [comment]
The version of Node in Electron 3 does not appear to maintain ABI compatibility with the version in Electron 4 despite both claiming to provide ABI version 64.

When you run `electron-rebuild` it calls `prebuild-install`, which correctly reports that it has a prebuilt binary for Electron with Node ABI version 64. However this binary fails with Electron 4 due to a change in Node ABI without a corresponding version increment.

You can hack your way around this by removing `prebuild-install` before running `electron-rebuild`.

```sh
npm install
rm -rf node_modules/prebuild-install
./node_modules/.bin/electron-rebuild
```

In summary, Electron needs to sort out its ABI versioning.

**Comment by lovell** [comment]
Here's confirmation of the Electron 3 vs 4 ABI incompatibility - https://github.com/lgeiger/node-abi/pull/55#issuecomment-452588421 - this will impact all native modules that provide prebuilt binaries for Electron.

**Comment by iacovlev-pavel** [comment]
@lovell Thank you for clarification and for such an awesome library in general.

**Comment by midevnull** [comment]
https://github.com/electron/electron/pull/16687 Am I correct that this should fix the issue?
But I am still getting an error when building with electron-builder on mac.

> pp threw an error during load
Error: The module '/*/node_modules/sharp/build/Release/sharp.node'
was compiled against a different Node.js version using
NODE_MODULE_VERSION 64. This version of Node.js requires
NODE_MODULE_VERSION 69. Please try re-compiling or re-installing
the module (for instance, using `npm rebuild` or `npm install`).
   at process.module.(anonymous function) [as dlopen] (ELECTRON_ASAR.js:160:31)
   at Object.Module._extensions..node (internal/modules/cjs/loader.js:722:18)
   at Object.module.(anonymous function) [as .node] (ELECTRON_ASAR.js:160:31)
   at Module.load (internal/modules/cjs/loader.js:602:32)
   at tryModuleLoad (internal/modules/cjs/loader.js:541:12)
   at Function.Module._load (internal/modules/cjs/loader.js:533:3)
   at Module.require (internal/modules/cjs/loader.js:640:17)
   at require (internal/modules/cjs/helpers.js:20:18)
   at bindings (/*/node_modules/sharp/node_modules/bindings/bindings.js:112:48)
   at Object.<anonymous> (/*/node_modules/sharp/lib/constructor.js:10:34)

This is probably a electron-builder issue and not a sharp one.

**Comment by lovell** [comment]
@midevnull https://github.com/electron-userland/electron-builder/issues/3660

**Comment by Cascoid** [comment]
Has the "Error: The specified module could not be found" had a proper solution yet?

Application is having the "Error: The specified module could not be found." issue on a windows machine.
`App threw an error during load
Error: The specified module could not be found.
\\?\C:\Users\Cascoid\Documents\rollapp\node_modules\sharp\build\Release\sharp.node
    at process.module.(anonymous function) [as dlopen] (ELECTRON_ASAR.js:160:31)
    at Object.Module._extensions..node (internal/modules/cjs/loader.js:722:18)
    at Object.module.(anonymous function) [as .node] (ELECTRON_ASAR.js:160:31)
    at Module.load (internal/modules/cjs/loader.js:602:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:541:12)
    at Function.Module._load (internal/modules/cjs/loader.js:533:3)
    at Module.require (internal/modules/cjs/loader.js:640:17)
    at require (internal/modules/cjs/helpers.js:20:18)
    at bindings (C:\Users\a707710\Documents\HOME\confetticode\TriviaRoll\node_modules\bindings\bindings.js:112:48)
    at Object.<anonymous> (C:\Users\Cascoid\Documents\rollapp\node_modules\sharp\lib\constructor.js:10:34)`
The application runs perfectly fine on mac.

 I've tried deleting node_modules and package-lock.json.. and reinstalling all node packages individually, still nothing.

Tried removing prebuild-install before rebuilding, also didn't work.

All other threads with similar issue point here, but not seeing a solution.

**Comment by xland** [comment]
Electron 4.1.0 seems to have solved this issue, is it?

> Electron 4.1.0 was released to work around issues users were experiencing with node-pre-gyp based native modules and our ABI number in Electron >= 4.0.4.
- 
> Fixed a crash when calling ECDH.setPrivateKey(). #17297
> TouchBarSegmentedControl instances now correctly update when you update the segments array and remove label or image dynamically. #17335
> https://electronjs.org/releases/stable#4.1.0

**Comment by tomaschyly** [comment]
I tried just Electron v4.1.0, but in my case it did not help. Gonna have to wait for sharp update.

**Comment by mceachen** [comment]
> I tried just Electron v4.1.0, but in my case it did not help. Gonna have to wait for sharp update.

Did you clean your build machine? I had to not only delete my project's node_modules, but 

`rm -rf ~/.electron ~/.electron-gyp ~/.npm/_libvips* ~/.node-gyp ~/.cache/yarn/*/*sharp*`

(I added it as postinstall script)

**Comment by tomaschyly** [comment]
Yes I did try to clean everything, but it seems that I will have to wait for update like I have been told here https://github.com/lovell/sharp/issues/1608

**Comment by lovell** [comment]
The ABI version of 69 used by Electron 4 is (correctly) considered "oddball" by the prebuild infrastructure so this is still blocked until the approach proposed in https://github.com/prebuild/prebuild-ci/issues/11 is addressed.

**Comment by JakeRadMSFT** [comment]
What is the workaround for this?

**Comment by xland** [comment]
@JakeRadMSFT 

Yes, I have rolled back to Electron 3.1.6(The last version of 3.x.x)

**Comment by JakeRadMSFT** [comment]
@xland have you gotten it to work after packaging it?

With 3.1.6, I'm hitting this issue: https://github.com/lovell/sharp/issues/1519.

**Comment by xland** [comment]
@JakeRadMSFT 
Electron 3.1.6 and sharp 0.21.3 work well together on macOS, Win10 and Ubuntu

[https://github.com/xland/xiangxuema](https://github.com/xland/xiangxuema)

**Comment by inukshuk** [comment]
I think everything already works with Electron 4.1.

I originally run into this same issue, but then I realized that `npm update --depth 9999` did not update sharp's sub-dependencies and I was stuck on older versions of `prebuild-install` and `node-abi`. Having updated those and also `node-gyp` globally sharp 0.21.3 works for me on Electron 4.1.

![Screenshot from 2019-03-22 09-39-58](https://user-images.githubusercontent.com/325102/54810615-d090f400-4c86-11e9-99da-a854c2a0ad8e.png)

I'm having unrelated issues with 0.22 / libvips 8.7.4 on Ubuntu Xenial (because of older versions of librsvg or cairo, not sure yet) but on newer Linux systems I had that combination working as well.

**Comment by JakeRadMSFT** [comment]
Fixed my issue on Mac (Mac works with Electron 4.1).

One windows, I still get "Error: The specified module could not be found". 

Now trying a downgrade to 3.1.6.

**Comment by JakeRadMSFT** [comment]
Sigh ... it's because of this: https://github.com/mapbox/node-pre-gyp/pull/279

**Comment by lovell** [comment]
Should anyone want to speed all this along, the best thing to do would be to work on https://github.com/prebuild/prebuild-ci/issues/11 and ensure ABI version 69 is added to the "oddball" list.

**Comment by JakeRadMSFT** [comment]
That's actually not the issue anymore ... I think it works at 4. I'll open a new bug ...

**Comment by barakman** [comment]
Encountered a similar issue with `web3` and `electron`.

Posted an MVP at https://github.com/electron/electron-quick-start/issues/269.

**Comment by lovell** [comment]
I've opened a PR to help ensure prebuilt binaries for the latest Electron 4 / ABI 69 will soon be available - see https://github.com/prebuild/prebuild-ci/pull/16

**Comment by giterium** [comment]
With Electron 3.1.8 output this
```
Uncaught Error: Could not locate the bindings file. Tried:
 → D:\ElectronApps\p360\node_modules\sharp\build\sharp.node
 → D:\ElectronApps\p360\node_modules\sharp\build\Debug\sharp.node
 → D:\ElectronApps\p360\node_modules\sharp\build\Release\sharp.node
 → D:\ElectronApps\p360\node_modules\sharp\out\Debug\sharp.node
 → D:\ElectronApps\p360\node_modules\sharp\Debug\sharp.node
 → D:\ElectronApps\p360\node_modules\sharp\out\Release\sharp.node
 → D:\ElectronApps\p360\node_modules\sharp\Release\sharp.node
 → D:\ElectronApps\p360\node_modules\sharp\build\default\sharp.node
 → D:\ElectronApps\p360\node_modules\sharp\compiled\10.2.0\win32\x64\sharp.node
 → D:\ElectronApps\p360\node_modules\sharp\addon-build\release\install-root\sharp.node
 → D:\ElectronApps\p360\node_modules\sharp\addon-build\debug\install-root\sharp.node
 → D:\ElectronApps\p360\node_modules\sharp\addon-build\default\install-root\sharp.node
 → D:\ElectronApps\p360\node_modules\sharp\lib\binding\node-v64-win32-x64\sharp.node
    at bindings (D:\ElectronApps\p360\node_modules\bindings\bindings.js:126)
    at Object.<anonymous> (D:\ElectronApps\p360\node_modules\sharp\lib\constructor.js:10)
    at Object.<anonymous> (D:\ElectronApps\p360\node_modules\sharp\lib\constructor.js:243)
    at Module._compile (internal/modules/cjs/loader.js:711)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:722)
    at Module.load (internal/modules/cjs/loader.js:620)
    at tryModuleLoad (internal/modules/cjs/loader.js:559)
    at Function.Module._load (internal/modules/cjs/loader.js:551)
    at Module.require (internal/modules/cjs/loader.js:658)
    at require (internal/modules/cjs/helpers.js:20)
```

**Comment by lovell** [comment]
v0.22.1 now available with prebuilt binaries for Electron 4 *and* 5.

Please remember to run `electron-rebuild` (which may itself also need updating to pick up the new ABI versions),

**Comment by xland** [comment]
This issue occur on electron 6.0.1.
@lovell 

```
electron-rebuild -f -w sharp

- Searching dependency tree
An unhandled error occurred inside electron-rebuild
Could not detect abi for version 6.0.1 and runtime electron.  Updating "node-abi" might help solve this issue if it is a new release of electron

Error: Could not detect abi for version 6.0.1 and runtime electron.  Updating "node-abi" might help solve this issue if it is a new release of electron
    at Object.getAbi (D:\project\xiangxuema\node_modules\node-abi\index.js:30:9)
    at new Rebuilder (D:\project\xiangxuema\node_modules\electron-rebuild\lib\src\rebuild.js:122:28)
    at rebuildWithOptions (D:\project\xiangxuema\node_modules\electron-rebuild\lib\src\rebuild.js:393:23)
    at Object.doRebuild [as rebuild] (D:\project\xiangxuema\node_modules\electron-rebuild\lib\src\rebuild.js:400:16)
    at Object.<anonymous> (D:\project\xiangxuema\node_modules\electron-rebuild\lib\src\cli.js:115:33)
    at Generator.next (<anonymous>)
    at fulfilled (D:\project\xiangxuema\node_modules\electron-rebuild\lib\src\cli.js:5:58)
npm ERR! code ELIFECYCLE
npm ERR! errno 4294967295
npm ERR! xiangxuema@6.4.3 buildsharp: `electron-rebuild -f -w sharp`
npm ERR! Exit status 4294967295
npm ERR!
npm ERR! Failed at the xiangxuema@6.4.3 buildsharp script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\allen\AppData\Roaming\npm-cache\_logs\2019-08-08T14_07_46_362Z-debug.log
```
